org.eclipse.xtext/src/org/eclipse/xtext/validation/DefaultUniqueNameContext.java

```diff
@@ -0,0 +1,319 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Sebastian Zarnekow and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.validation;
+
+import java.util.Collection;
+import java.util.List;
+
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EcorePackage;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.xtext.resource.IContainer;
+import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.IResourceDescriptionsProvider;
+import org.eclipse.xtext.resource.IResourceServiceProvider;
+import org.eclipse.xtext.resource.ISelectable;
+import org.eclipse.xtext.resource.impl.AbstractCompoundSelectable;
+import org.eclipse.xtext.scoping.ICaseInsensitivityHelper;
+import org.eclipse.xtext.scoping.impl.CaseInsensitivityHelper;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.validation.INamesAreUniqueValidationHelper.Context;
+
+import com.google.common.collect.Iterables;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * Standard implementation of a {@link INamesAreUniqueValidationHelper.Context}.
+ * 
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.22
+ */
+public class DefaultUniqueNameContext implements INamesAreUniqueValidationHelper.Context {
+
+	/**
+	 * Base class for {@link INamesAreUniqueValidationHelper.ContextProvider}.
+	 */
+	public static abstract class BaseContextProvider implements INamesAreUniqueValidationHelper.ContextProvider {
+		@Inject
+		private IResourceServiceProvider.Registry resourceServiceProviderRegistry = IResourceServiceProvider.Registry.INSTANCE;
+		@Inject
+		private ICaseInsensitivityHelper caseInsensitivityHelper = new CaseInsensitivityHelper();
```

Comment:
Is the assignment necessary here? The interface uses `@ImplementedBy(CaseInsensitivityHelper.class)`

Reply:
I tried to make this as failsafe as I could think of. The org.eclipse.xtext.validation.NamesAreUniqueValidator has setters for its injected fields. That means, it was possible to use it without an injector - so did our tests. Now it has an additional field of type `INamesAreUniqueValidationHelper.ContextProvider` which may not be null by default to not break things. If this was on a green field, I'd probably used constructor injection.